saml/src/main/java/com/linecorp/armeria/saml/JwtBasedSamlRequestIdManager.java

```diff
@@ -0,0 +1,101 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.saml;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.UnsupportedEncodingException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.JWTVerifier;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.google.common.base.MoreObjects;
+
+import io.netty.util.internal.ThreadLocalRandom;
+
+/**
+ * A {@link SamlRequestIdManager} implementation based on JSON Web Tokens specification.
+ * See <a href="https://jwt.io/">JSON Web Tokens</a> for more information.
+ */
+final class JwtBasedSamlRequestIdManager implements SamlRequestIdManager {
+
+    private static final String CLAIM_NAME_SALT = "salt";
+
+    private final String issuer;
+    private final int validSeconds;
+    private final int leewaySeconds;
+
+    private final Algorithm algorithm;
+    private final JWTVerifier verifier;
+
+    JwtBasedSamlRequestIdManager(String issuer, String secret, int validSeconds, int leewaySeconds) {
+        this.issuer = requireNonNull(issuer, "issuer");
+        this.validSeconds = validSeconds;
+        this.leewaySeconds = leewaySeconds;
+
+        checkArgument(validSeconds > 0,
+                      "invalid valid duration: " + validSeconds + " (expected: > 0)");
+        checkArgument(leewaySeconds >= 0,
+                      "invalid leeway duration:" + leewaySeconds + " (expected: >= 0)");
+        try {
+            algorithm = Algorithm.HMAC384(requireNonNull(secret, "secret"));
+            verifier = JWT.require(algorithm)
+                          .withIssuer(issuer)
+                          .acceptLeeway(leewaySeconds)
+                          .build();
+        } catch (UnsupportedEncodingException e) {
+            throw new Error("cannot instantiate " + Algorithm.class.getSimpleName() + " instance", e);
+        }
+    }
+
+    @Override
+    public String generateNewId() {
+        final Instant now = Instant.now();
+        return JWT.create()
+                  .withIssuer(issuer)
+                  .withIssuedAt(Date.from(now))
+                  .withExpiresAt(Date.from(now.plus(validSeconds, ChronoUnit.SECONDS)))
+                  .withClaim(CLAIM_NAME_SALT, ThreadLocalRandom.current().nextInt(Integer.MAX_VALUE))
+                  .sign(algorithm);
+    }
+
+    @Override
+    public boolean validateId(String id) {
+        requireNonNull(id, "id");
+        try {
+            // Verifier will check whether its issuer is me and also check whether it has been expired or not.
+            final DecodedJWT jwt = verifier.verify(id);
+            return !jwt.getClaim(CLAIM_NAME_SALT).isNull();
```

Comment:
What does this salt do? If we don't actually compare it to anything, it doesn't seem to add any checking.

Reply:
As I said [here](https://github.com/line/armeria/pull/1292#discussion_r203948109), JWT library is based on `seconds`. It means that it generates same IDs in a seconds because we have the same `issuer` and time values in a second. So the `salt` makes the ID different for every time. 

We can do that as simply adding a sequentially increasing number as a claim. Is it better? Let me know what you think. :-)